posit-dev/positron
# Intrinsic plot sizing

This PR adds a new sizing policy, the intrinsic sizing policy, which renders the plot at its "intrinsic" size i.e. the original size of the plot known to the backend.

Addresses https://github.com/posit-dev/positron/issues/4040.

Here's the corresponding Ark PR: https://github.com/posit-dev/ark/pull/470.

### How it works

1. Creating a plot with a backend that supports this feature, will add a new sizing policy. For example, in Python:

    ```python
    import matplotlib.pyplot as plt
    plt.subplots(figsize=(3, 10))[1].plot([0, 1])
    ```

2. The sizing policy's label will contain the source of the sizing information and the original size in the backend's preferred unit of measurement:

    <img width="218" alt="image" src="https://github.com/user-attachments/assets/71e50c9a-0caf-4f5b-bc5b-37db3e246060">

3. You can also save plots at their intrinsic size:
    
    <img width="503" alt="image" src="https://github.com/user-attachments/assets/92c1773c-3533-44ad-98e0-ce14737dac93">

4. If the backend does not support intrinsic sizing, the policy's label will be disabled:

    <img width="432" alt="image" src="https://github.com/user-attachments/assets/17babb0a-8012-4939-8602-e0b8dd756021">

    1. If you switch from a plot with the intrinsic sizing policy selected to a plot that does not support intrinsic sizing, it will fall back to the "auto" sizing policy.
    2. The "Use intrinsic sizing" checkbox will also be disabled in the save modal.

### QA Notes

1. Create Python plots with different sizes and make sure that switching between them updates the plot size and the sizing policy's label, e.g:
    ```python
    import matplotlib.pyplot as plt
    plt.subplots(figsize=(10, 3))[1].plot([0, 1])
    plt.subplots(figsize=(3, 10))[1].plot([0, 1])
    plt.subplots(figsize=(3, 4))[1].plot([0, 1])
    plt.subplots(figsize=(4, 3))[1].plot([0, 1])
    plt.subplots()[1].plot([0, 1])
    ```
2. Switching from a Python plot with the intrinsic sizing policy selected to an R plot should reset to the auto sizing policy.
